nixos/modules/services/networking/inadyn.nix

```diff
@@ -0,0 +1,193 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.inadyn;
+
+  # check if a value of an attrset is not null or an empty collection
+  nonEmptyValue = _: v: ! isNull v && v != [ ] && v != { };
+
+  renderOption = k: v:
+    if builtins.elem k [ "provider" "custom" ] then
+      lib.concatStringsSep "\n"
+        (mapAttrsToList
+          (name: config: ''
+            ${k} ${name} {
+                ${lib.concatStringsSep "\n    " (mapAttrsToList renderOption (filterAttrs nonEmptyValue config))}
+            }'')
+          v)
+    else if k == "include" then
+      "${k}(\"${v}\")"
+    else if k == "hostname" && builtins.isList v then
+      "${k} = { ${builtins.concatStringsSep ", " (map (s: "\"${s}\"") v)} }"
+    else if builtins.isBool v then
+      "${k} = ${boolToString v}"
+    else if builtins.isString v then
+      "${k} = \"${v}\""
+    else
+      "${k} = ${toString v}";
+
+  configFile' = pkgs.writeText "inadyn.conf"
+    ''
+    # This file was generated by nix
+    # do not edit
+
+    ${(lib.concatStringsSep "\n" (mapAttrsToList renderOption (filterAttrs nonEmptyValue cfg.settings)))}
+    '';
+
+  configFile = if (cfg.configFile != null) then cfg.configFile else configFile';
+in
+{
+  options.services.inadyn = with types;
+    let providerOptions =
+      {
+        include = mkOption {
+          default = null;
+          description = mdDoc "File to include additional settings for this provider from.";
+          type = nullOr path;
+        };
+        ssl = mkOption {
+          default = true;
+          description = mdDoc "Whether to use HTTPS for this DDNS provider.";
+          type = bool;
+        };
+        username = mkOption {
+          default = null;
+          description = mdDoc "Username for this DDNS provider.";
+          type = nullOr str;
+        };
+        password = mkOption {
```

Comment:
I'm not sure if it's a blocker, but personally I can't/won't use this module before it has a passwordFile option that doesn't leak secrets.

Btw, here's another inadyn NixOS module I saw recently, in case it can provide some inspiration: https://github.com/Mic92/sops-nix/issues/412#issue-1920132366.

Reply:
> To store credentials securely, use the `include` or `configFile` options.

I saw this 5 seconds too late. Can you add an example for `include`? Or add an `passwordFile` option? :smile:

Reply:
Yes, a couple of useful ideas in that useful module – thanks for pointing it out!

As for protecting your credentials, the "nix-y" way would to be something like
```nix
  services.inadyn = {
    enable = true;
    settings = {
      custom."namecheap" = {
        include     = ./auth.txt;
        ddns-server = "dynamicdns.park-your-domain.com";
        ddns-path   = "/update?domain=%u&password=%p&host=%h&ip=%i";
        hostname    = [ "@" "*" ];
      };
    };
  };
```

where `auth.txt` contains your user/password (and any other options that would usually belong in a custom/provider block, e.g.:
```
user = "USERNAME"
password = "PASSWORD"
```

Note that this is functionality built-in to inadyn, not custom logic for this module.

You can also, of course, just use `configFile`.

Reply:
Can we educate people on this?